config/src/inline/mod.rs

```diff
@@ -0,0 +1,141 @@
+where
+    T: ConfParser + 'static,
+    P: AsRef<Path>,
+{
+    type Error = ConfParserError;
+
+    /// Tries to create an instance of `Self`, detecting inline configurations from the project
+    /// compile output.
+    ///
+    /// Param is a tuple, whose elements are:
+    /// 1. Solidity compiler output, essential to extract comments from compiled functions.
+    /// 2. A reference to a base configuration. This essentially works as a fallback.
+    /// 3. A root path to express contract base dirs. This is essential to match inline configs at
+    /// runtime.
+    fn try_from(value: (&'a ProjectCompileOutput, &'a T, &'a P)) -> Result<Self, Self::Error> {
+        let mut configs: HashMap<(String, String), T> = HashMap::new();
+        let compile_output = value.0.clone();
+        let base_conf = value.1;
+        let root = value.2;
+
+        for artifact in compile_output.with_stripped_file_prefixes(root).into_artifacts() {
+            if let Some(ast) = artifact.1.ast.as_ref() {
+                let contract_id: String = artifact.0.identifier();
+                if let Some(node) = contract_root_node(&ast.nodes, &contract_id) {
+                    try_apply(base_conf, &mut configs, &contract_id, node)?;
+                }
+            }
+        }
+
+        Ok(Self { configs })
+    }
+}
+
+/// Given a list of nodes, find a "ContractDefinition" node that matches
+/// the provided contract_id.
+fn contract_root_node<'a>(nodes: &'a [Node], contract_id: &'a str) -> Option<&'a Node> {
+    for n in nodes.iter() {
+        if format!("{:?}", n.node_type) == *"ContractDefinition" {
+            let contract_data = &n.other;
+            if let Value::String(contract_name) = contract_data.get("name")? {
+                if contract_id.ends_with(contract_name) {
+                    return Some(n)
+                }
+            }
+        }
+    }
+    None
+}
+
+/// Implements a DFS over a compiler output node and its children.
+/// If a configuration is found for a solidity function, it is added to
+/// `map` under the (contract id, function name) key.
+/// This function may result in parsing errors (see [`ConfParserError`]).
+fn try_apply<T>(
+    base_conf: &T,
+    map: &mut HashMap<(String, String), T>,
+    contract_id: &str,
+    node: &Node,
+) -> Result<(), ConfParserError>
+where
+    T: ConfParser + 'static,
+{
+    for n in node.nodes.iter() {
+        if let Some((fn_name, fn_docs)) = get_fn_data(n) {
+            if let Some(conf) = base_conf.try_merge(fn_docs.clone())? {
+                // We found a config that applies for the pair contract_id, fn_name.
+                let key = (contract_id.into(), fn_name);
+                map.insert(key, conf);
+            }
+        }
+
+        try_apply(base_conf, map, contract_id, n)?;
+    }
+    Ok(())
+}
+
+fn get_fn_data(node: &Node) -> Option<(String, String)> {
+    if format!("{:?}", node.node_type) == *"FunctionDefinition" {
```

Comment:
hmm, isn't this an enum?
can we use the `matches!` macro here?

Reply:
You are right Matt, but It looks like the enum definition is not publicly exported
`ethers_solc::artifacts::ast::lowfidelity::NodeType`

Any advice on this?

Reply:
hmm, it should be exported here:

https://github.com/gakonst/ethers-rs/blob/main/ethers-solc/src/artifacts/ast/mod.rs#L22-L22

Reply:
Way better ty